### PR TITLE
Web3signer keymanager: key manager implementation 

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**
Feature

Part of #9994

**What does this PR do? Why is it needed?**
This PR is the second in a series that aims to add support for Consensys' Web3Signer remote signing server, supported by various eth2 clients today. Web3signer is an important standard for institutional adoption of staking, and we should support it in Prysm today.

This pr defines the remote web3signer keymanager object that follows the keymanager interface. this will allow other parts of prysm to interface with the web3 signer. This key manager object is not yet used by the rest of Prysm, but will be in subsequent PRs related to this series.

### CODE REVIEW

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+		GenesisValidatorsRoot: hexutil.Encode(km.genesisValidatorsRoot),
+	}
+	aggregationSlotData := &AggregationSlot{Slot: fmt.Sprint(request.AggregationSlot)}
+	web3SignerRequest := SignRequest{
+		Type:            signRequestType,
+		ForkInfo:        forkInfoData,
+		SigningRoot:     hexutil.Encode(request.SigningRoot),
+		AggregationSlot: aggregationSlotData,
+	}
+	return km.client.Sign(string(request.PublicKey), &web3SignerRequest)
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/keymanager.go:120]: Use hexutil.Encode

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+	case *validatorpb.SignRequest_ContributionAndProof:
+		return "SYNC_COMMITTEE_CONTRIBUTION_AND_PROOF", nil
+	default:
+		return "", errors.New("Web3signer sign request type not found")
+	}
+}
+
+// SubscribeAccountChanges returns the event subscription for changes to public keys
+func (km *Keymanager) SubscribeAccountChanges(pubKeysChan chan [][48]byte) event.Subscription {
+	// not used right now
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/keymanager.go:160]: Explain why this is a stub. Current comment is not very helpful to the reader

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+	return event.NewSubscription(func(i <-chan struct{}) error {
+		return nil
+	})
+}
+
+// reloadKeys reloads the public keys from the remote server
+func (km *Keymanager) reloadKeys() {
+	// not used right now
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/keymanager.go:168]: Explain why this is a stub. Current comment is not very helpful to the reader

File: validator/keymanager/remote-web3signer/keymanager_test.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+	validatorpb "github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1/validator-client"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/prysmaticlabs/prysm/encoding/bytesutil"
+	"golang.org/x/net/context"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/prysmaticlabs/prysm/crypto/bls"
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/keymanager_test.go:19]: goimports and grouping here

File: validator/keymanager/remote-web3signer/keymanager_test.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	resp, err := km.FetchValidatingPublicKeys(ctx)
+	if err != nil {
+		fmt.Printf("error: %v", err)
+	}
+	assert.NotNil(t, resp)
+	assert.Nil(t, err)
+	assert.EqualValues(t, resp, keys)
+}
+
+//TODO: not a very useful test as we aren't looking at reload for keymanager jsut yet
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/keymanager_test.go:152]: Would delete in that case until we do need tests for it

File: go.mod [hunk 6]
```diff
@@ -105,6 +105,7 @@ require (
 	github.com/go-ole/go-ole v1.2.5 // indirect
 	github.com/peterh/liner v1.2.0 // indirect
 	github.com/prometheus/tsdb v0.10.0 // indirect
+	golang.org/x/net v0.0.0-20211112202133-69e39bad7dc2
```
Comment (rauljordan) [inline go.mod:108]: Revert

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+	"github.com/prysmaticlabs/prysm/async/event"
+	"github.com/prysmaticlabs/prysm/crypto/bls"
+	validatorpb "github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1/validator-client"
+	"github.com/prysmaticlabs/prysm/validator/keymanager"
+)
+
+// Web3SignerKeyManager interface implements Ikeymanager interface
+type Web3SignerKeyManager interface {
+	keymanager.IKeymanager
+}
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply validator/keymanager/remote-web3signer/keymanager.go:20]: will update to include the other properties forgot to go back to it

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+	}
+}
+
+// SetupConfig includes configuration values for initializing
+// a keymanager, such as passwords, the wallet, and more.
+type SetupConfig struct {
+	Option                *PublicKeysOption
+	BaseEndpoint          string
+	GenesisValidatorsRoot []byte
+}
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply validator/keymanager/remote-web3signer/keymanager.go:47]: all the other keymanagers use the setup config, but good point I'll just change it to use an array of options

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+	if km.publicKeysURL != "" {
+		return km.client.GetPublicKeys(km.publicKeysURL)
+	}
+	return km.providedPublicKeys, nil
+}
+
+// Sign signs the message by using a remote web3signer server
+func (km *Keymanager) Sign(ctx context.Context, request *validatorpb.SignRequest) (bls.Signature, error) {
+
+	if request.Fork == nil {
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply validator/keymanager/remote-web3signer/keymanager.go:92]: It does

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 10]
```diff
@@ -0,0 +1,169 @@
+package remote_web3signer
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+
+	"github.com/pkg/errors"
+
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply validator/keymanager/remote-web3signer/keymanager.go]: I believe I have this 🤔  we set this up  must have happened when determining that the auto-import was incorrect

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+func (km *Keymanager) Sign(ctx context.Context, request *validatorpb.SignRequest) (bls.Signature, error) {
+
+	if request.Fork == nil {
+		return nil, errors.New("invalid sign request: Fork is nil")
+	}
+	if request.AggregationSlot == 0 {
+		return nil, errors.New("invalid sign request: AggregationSlot is 0")
+	}
+
+	// get new keys before signing
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply validator/keymanager/remote-web3signer/keymanager.go:83]: will update

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+	return km.providedPublicKeys, nil
+}
+
+// Sign signs the message by using a remote web3signer server
+func (km *Keymanager) Sign(ctx context.Context, request *validatorpb.SignRequest) (bls.Signature, error) {
+
+	if request.Fork == nil {
+		return nil, errors.New("invalid sign request: Fork is nil")
+	}
+	if request.AggregationSlot == 0 {
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply validator/keymanager/remote-web3signer/keymanager.go]: removed for now asking about the genesis bool

File: validator/keymanager/remote-web3signer/client.go [hunk 13]
```diff
@@ -2,34 +2,41 @@ package remote_web3signer
...
 	ethApiNamespace = "/api/v1/eth2/sign/"
 	maxTimeout      = 3 * time.Second
 )
 
+// Web3SignerClient defines the interface for interacting with a remote web3signer.
+type Web3SignerClient interface {
+	Sign(pubKey string, request *SignRequest) (bls.Signature, error)
+	GetPublicKeys(url string) ([][48]byte, error)
+	//ReloadSignerKeys() error
+	//GetServerStatus() (string, error)
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/client.go:28]: Removed commented code. If needed, we can add it later

We should also rename this to be more clear about what it does
`httpSignerClient interface` perhaps. It should probably also be unexported

File: validator/keymanager/remote-web3signer/client.go [hunk 14]
```diff
@@ -93,33 +100,32 @@ func (client *client) Sign(pubKey string, request *SignRequest) (bls.Signature,
...
-	decoded, err := decodeHex(signResp.Signature)
+	decoded, err := hexutil.Decode(signResp.Signature)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "failed to decode signature")
 	}
 	return bls.SignatureFromBytes(decoded)
 }
 
 // GetPublicKeys is a wrapper method around the web3signer publickeys api (this may be removed in the future or moved to another location due to its usage).
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/client.go:110]:
```suggestion
// GetPublicKeys is a wrapper method around the web3signer publickeys api.
```

File: validator/keymanager/remote-web3signer/client.go [hunk 15]
```diff
@@ -93,33 +100,32 @@ func (client *client) Sign(pubKey string, request *SignRequest) (bls.Signature,
...
-func (client *client) GetPublicKeys() ([][]byte, error) {
-	const requestPath = "/publicKeys"
-	resp, err := client.doRequest(http.MethodGet, client.BasePath+requestPath, nil)
+func (client *client) GetPublicKeys(url string) ([][48]byte, error) {
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/client.go:111]: Most methods should take in a context as a first argument `ctx context.Context, url string`, as it is a common pattern in the codebase

File: validator/keymanager/remote-web3signer/client.go [hunk 16]
```diff
@@ -93,33 +100,32 @@ func (client *client) Sign(pubKey string, request *SignRequest) (bls.Signature,
...
+	resp, err := client.doRequest(http.MethodGet, url, nil)
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/client.go:112]:
```suggestion
	resp, err := client.doRequest(http.MethodGet, url, nil /* add explanation for what the nil is */)
```

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+)
+
+// SetupConfig includes configuration values for initializing.
+// a keymanager, such as passwords, the wallet, and more.
+// Web3Signer contains one public keys option. Either through a URL or a static key list.
+type SetupConfig struct {
+	BaseEndpoint          string
+	GenesisValidatorsRoot []byte
+
+	//Either URL or keylist must be set.
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/keymanager.go:21]:
```suggestion
	// Either URL or keylist must be set.
```

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+	// If the URL is set, the keymanager will fetch the public keys from the URL.
+	// caution: this option is susceptible to slashing if the web3signer's validator keys are shared across validators
+	PublicKeysURL string
+
+	//Either URL or keylist must be set.
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/keymanager.go:26]:
```suggestion
	// Either URL or keylist must be set.
```

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+	//a static list of public keys to be passed by the user to determine what accounts should sign.
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/keymanager.go:27]:
```suggestion
	// a static list of public keys to be passed by the user to determine what accounts should sign.
```

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+	// This will provide a layer of safety against slashing if the web3signer is shared across validators.
+	ProvidedPublicKeys [][48]byte
+}
+
+// Keymanager defines the web3signer keymanager.
+type Keymanager struct {
+	client                Web3SignerClient
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/keymanager.go:34]: rename to apiClient

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+	if cfg.BaseEndpoint == "" || len(cfg.GenesisValidatorsRoot) == 0 {
+		return nil, errors.New("invalid setup config, one or more configs are empty: " + fmt.Sprintf("BaseEndpoint: %v, GenesisValidatorsRoot: %v.", cfg.BaseEndpoint, cfg.GenesisValidatorsRoot))
+	}
+	if cfg.PublicKeysURL != "" && len(cfg.ProvidedPublicKeys) != 0 {
+		return nil, errors.New("Either a provided list of public keys or a URL to a list of public keys must be provided, but not both")
+	}
+	if cfg.PublicKeysURL == "" && len(cfg.ProvidedPublicKeys) == 0 {
+		return nil, errors.New("no valid public key options provided")
+	}
+	client, err := newClient(cfg.BaseEndpoint)
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/keymanager.go:52]: I would call this function `newApiClient`

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+			return nil, err
+		}
+		km.providedPublicKeys = providedPublicKeys
+	}
+	return km.providedPublicKeys, nil
+}
+
+// Sign signs the message by using a remote web3signer server.
+func (km *Keymanager) Sign(_ context.Context, request *validatorpb.SignRequest) (bls.Signature, error) {
+
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/keymanager.go:79]:
```suggestion
```

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+		GenesisValidatorsRoot: hexutil.Encode(km.genesisValidatorsRoot),
+	}
+	aggregationSlotData := &AggregationSlot{Slot: fmt.Sprint(request.AggregationSlot)}
+	web3SignerRequest := SignRequest{
+		Type:            signRequestType,
+		ForkInfo:        forkInfoData,
+		SigningRoot:     hexutil.Encode(request.SigningRoot),
+		AggregationSlot: aggregationSlotData,
+	}
+	return km.client.Sign(hexutil.Encode(request.PublicKey), &web3SignerRequest)
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/keymanager.go:105]: the client sign function should ideally take in a `ctx context.Context`

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+	/*
+		case *validatorpb.SignRequest_BlockV3:
+		return "BLOCK_V3", nil
+	*/
+
+	// We do not support "DEPOSIT" type.
+	/*
+		case *validatorpb.:
+		return "DEPOSIT", nil
+	*/
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/keymanager.go:131]: How will we remember to do this? This is easy to forget. Maybe we can have an open issue and then add the comment
```
// TODO(#10020): Need to add support for merge blocks
```

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 25]
```diff
@@ -0,0 +1,162 @@
...
+	case *validatorpb.SignRequest_Epoch:
+		return "RANDAO_REVEAL", nil
+	case *validatorpb.SignRequest_Exit: //not sure
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/keymanager.go:134]:
```suggestion
	case *validatorpb.SignRequest_Exit:
```

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 26]
```diff
@@ -0,0 +1,162 @@
...
+		return "VOLUNTARY_EXIT", nil
+	case *validatorpb.SignRequest_SyncMessageBlockRoot:
+		return "SYNC_COMMITTEE_MESSAGE", nil
+	case *validatorpb.SignRequest_SyncAggregatorSelectionData:
+		return "SYNC_COMMITTEE_SELECTION_PROOF", nil
+	case *validatorpb.SignRequest_ContributionAndProof:
+		return "SYNC_COMMITTEE_CONTRIBUTION_AND_PROOF", nil
+	default:
+		return "", errors.New("Web3signer sign request type not found")
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/keymanager.go:143]: Maybe add to the error the kind of sign request that was passed in and not found

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 27]
```diff
@@ -0,0 +1,162 @@
...
+		return nil
+	})
+}
+
+// reloadKeys reloads the public keys from the remote server
+func (km *Keymanager) reloadKeys() {
+	// Not used right now.
+	// The feature of needing to dynamically reload from the validator instead of from the web3signer is yet to be determined.
+	// In the future there may be an api provided to add remote sign keys to the static list or remove from the static list.
+}
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/keymanager.go:162]:
```suggestion
```

File: validator/keymanager/remote-web3signer/client.go [hunk 28]
```diff
@@ -2,38 +2,44 @@ package remote_web3signer
...
 
-// client a wrapper object around web3signer APIs. API docs found here https://consensys.github.io/web3signer/web3signer-eth2.html.
-type client struct {
+// httpSignerClient defines the interface for interacting with a remote web3signer.
+type httpSignerClient interface {
+	Sign(ctx context.Context, pubKey string, request *SignRequest) (bls.Signature, error)
+	GetPublicKeys(ctx context.Context, url string) ([][48]byte, error)
+}
+
+// apiClient a wrapper object around web3signer APIs. API docs found here https://consensys.github.io/web3signer/web3signer-eth2.html.
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/client.go:30]: Note: unexported methods do not need to follow commenting conventions. Exported ones need to have the name of the thing as the first word, but unexported can be anything you prefer

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 29]
```diff
@@ -0,0 +1,154 @@
...
+	genesisValidatorsRoot []byte
+	publicKeysURL         string
+	providedPublicKeys    [][48]byte
+	accountsChangedFeed   *event.Feed
+}
+
+// NewKeymanager instantiates a new web3signer key manager.
+func NewKeymanager(_ context.Context, cfg *SetupConfig) (*Keymanager, error) {
+	if cfg.BaseEndpoint == "" || len(cfg.GenesisValidatorsRoot) == 0 {
+		return nil, errors.New("invalid setup config, one or more configs are empty: " + fmt.Sprintf("BaseEndpoint: %v, GenesisValidatorsRoot: %v.", cfg.BaseEndpoint, cfg.GenesisValidatorsRoot))
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/keymanager.go:44]: use fmt.Errorf instead of errors.New to get formatting abilities

**Review by rauljordan — APPROVED** [review]

**Review by rauljordan — APPROVED** [review]

**Pull request merged**